pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java

```diff
@@ -0,0 +1,360 @@
+import com.linkedin.pinot.common.utils.SchemaUtils;
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.net.URLConnection;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nonnull;
+import org.apache.helix.PropertyPathConfig;
+import org.apache.helix.PropertyType;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.manager.zk.ZKHelixAdmin;
+import org.apache.helix.manager.zk.ZNRecordSerializer;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The <code>AutoAddInvertedIndex</code> class can be used to automatically add inverted index to tables based on the
+ * strategy and mode specified.
+ * <p>Currently support:
+ * <ul>
+ *   <li>
+ *     Strategy:
+ *     <ul>
+ *       <li>QUERY: add inverted index based on the query result</li>
+ *     </ul>
+ *   </li>
+ *   <li>
+ *     Mode:
+ *     <ul>
+ *       <li>NEW: apply only to tables without inverted index</li>
+ *       <li>REMOVE: remove all auto-generated inverted index</li>
+ *       <li>REFRESH: refresh the auto-generated inverted index</li>
+ *       <li>APPEND: append to the auto-generated inverted index</li>
+ *     </ul>
+ *   </li>
+ * </ul>
+ */
+public class AutoAddInvertedIndex {
+  public enum Strategy {
+    QUERY     // Add inverted index based on the query result
+  }
+
+  public enum Mode {
+    NEW,      // Apply only to tables without inverted index
+    REMOVE,   // Remove all auto-generated inverted index
+    REFRESH,  // Refresh the auto-generated inverted index
+    APPEND    // Append to the auto-generated inverted index
+  }
+
+  public static final long DEFAULT_TABLE_SIZE_THRESHOLD = 10_000_000;
+  public static final long DEFAULT_CARDINALITY_THRESHOLD = 100;
+  public static final int DEFAULT_MAX_NUM_INVERTED_INDEX_ADDED = 2;
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AutoAddInvertedIndex.class);
+  private static final String TABLE_CONFIG_PATH = "/CONFIGS/TABLE/";
+  private static final String SCHEMA_PATH = "/SCHEMAS/";
+
+  private final String _clusterName;
+  private final String _controllerAddress;
+  private final String _brokerAddress;
+  private final ZKHelixAdmin _helixAdmin;
+  private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
+  private final Strategy _strategy;
+  private final Mode _mode;
+
+  private String _tableNamePattern = null;
+  private long _tableSizeThreshold = DEFAULT_TABLE_SIZE_THRESHOLD;
+  private long _cardinalityThreshold = DEFAULT_CARDINALITY_THRESHOLD;
+  private int _maxNumInvertedIndexAdded = DEFAULT_MAX_NUM_INVERTED_INDEX_ADDED;
+
+  public AutoAddInvertedIndex(@Nonnull String zkAddress, @Nonnull String clusterName, @Nonnull String controllerAddress,
```

Comment:
Because PinotHelixResourceManager might create a new controller, for now I will use ZKHelixAdmin. When we actually add this util into controller in the future, we can add a new constructor with PinotHelixResourceManager.